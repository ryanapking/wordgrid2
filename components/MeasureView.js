import React, { Component } from 'react';
import { View, ViewPropTypes } from 'react-native';
import PropTypes from 'prop-types';

export default class MeasureView extends Component {
  render() {
    return (
      <View
        style={this.props.style}
        ref={ measureView => this.measureView = measureView }
        onLayout={() => this._onLayout() }
      >
        {this.props.children}
      </View>
    );
  }

  _onLayout() {
    this.measureView.measure((x, y, width, height, pageX, pageY) => {
      this.props.onMeasure(x, y, width, height, pageX, pageY);
    });
  }

  static propTypes = {
    onMeasure: PropTypes.func.isRequired,
    style: ViewPropTypes.style,
  }
}