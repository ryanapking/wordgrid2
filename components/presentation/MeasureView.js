import React, { useState } from 'react';
import { View, ViewPropTypes } from 'react-native';
import PropTypes from 'prop-types';

const MeasureView = props => {
  const [viewToMeasure, setViewToMeasure] = useState(null);

  const measure = () => {
    viewToMeasure.measure((x, y, width, height, pageX, pageY) => {
      props.onMeasure(x, y, width, height, pageX, pageY);
    });
  };

  return (
    <View
      style={props.style}
      ref={ measureView => setViewToMeasure(measureView) }
      onLayout={() => measure() }
      { ...props.panHandlers }
      pointerEvents={ props.pointerEvents ? props.pointerEvents : 'auto' }
    >
      {props.children}
    </View>
  );
};

MeasureView.propTypes = {
  onMeasure: PropTypes.func.isRequired,
  pointerEvents: PropTypes.string,
  panHandlers: PropTypes.object,
  style: ViewPropTypes.style,
};

export default MeasureView;