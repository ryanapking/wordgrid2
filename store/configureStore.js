import { createStore, applyMiddleware } from 'redux';
import thunk from 'redux-thunk';

import reducer from '../ducks/reducer';

export default function configureStore(initialState) {
  return createStore(
    reducer,
    applyMiddleware(thunk)
  );
}