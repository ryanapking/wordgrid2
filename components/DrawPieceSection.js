import React, { useState, useEffect } from 'react';
import { StyleSheet, View, ViewPropTypes } from 'react-native';
import { useDispatch } from 'react-redux';
import PropTypes from 'prop-types';

import { rowsArrayType } from "../proptypes";
import PieceDraggableView from './PieceDraggableView';
import PieceDrawLetterGrid from "./PieceDrawLetterGrid";
import MeasureView from "./MeasureView";
import { setPieceLocation as reduxSetPieceLocation, clearPieceLocations } from "../data/redux/gameDisplay";

const DrawPieceSection = props => {
  const [pieceLocations, setPieceLocations] = useState({});
  const dispatch = useDispatch();
  const { allowDrag, pieces } = props;

  const setPieceLocation = (piece, pieceIndex, location) => {
    const pieceLocation = { ...location, piece };

    // set local pieceLocation
    setPieceLocations(pieceLocations => {
      return {
        ...pieceLocations,
        [pieceIndex]: pieceLocation
      }
    });

    // set the piece location for the overlay
    if (allowDrag) {
      dispatch(reduxSetPieceLocation(pieceIndex, pieceLocation));
    }
  };

  // clear piece locations from redux on unmount
  useEffect(() => {
    return () => {
      /* istanbul ignore else*/
      if (allowDrag) {
        dispatch(clearPieceLocations());
      }
    }
  }, []);

  return (
    <View style={[props.style, styles.container]}>
      { pieces.map( (piece, pieceIndex) => {
        const baseSize = pieceLocations[pieceIndex] ? pieceLocations[pieceIndex].width : 0;
        return (
          <View style={styles.gamePieceContainer} key={pieceIndex}>
            <MeasureView
              style={styles.gamePiece}
              onMeasure={ (x, y, width, height, pageX, pageY) => setPieceLocation(piece, pieceIndex, {x, y, width, height, pageX, pageY}) }
            >
              { allowDrag ? null :
                <PieceDraggableView
                  piece={piece}
                  pieceIndex={pieceIndex}
                  style={styles.gamePiece}
                  allowDrag={allowDrag}
                  baseSize={baseSize}
                >
                  <PieceDrawLetterGrid
                    piece={piece}
                    pieceSize={baseSize}
                  />
                </PieceDraggableView>
              }
            </MeasureView>
          </View>
        );
      })}
    </View>
  );
};

DrawPieceSection.propTypes = {
  allowDrag: PropTypes.bool,
  pieces: rowsArrayType.isRequired,
  style: ViewPropTypes.style,
};

const styles = StyleSheet.create({
  container: {
    display: 'flex',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'center',
    width: '100%',
    height: '100%',
  },
  gamePieceContainer: {
    flex: 1,
    margin: 2,
    display: 'flex',
    flexDirection: 'column',
    justifyContent: 'center',
    maxHeight: '100%',
    maxWidth: '100%',
  },
  gamePiece: {
    backgroundColor: 'gray',
    maxWidth: '100%',
    maxHeight: '100%',
    aspectRatio: 1,
  }
});

export default DrawPieceSection;