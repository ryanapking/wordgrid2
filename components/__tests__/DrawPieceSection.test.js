import React from 'react';
import { Provider } from "react-redux";
import { render, fireEvent } from "react-native-testing-library";

import MeasureView from "../MeasureView";
import DrawPieceSection from "../DrawPieceSection";

import configureStore from '../../data/redux/configureStore';
const store = configureStore();

const testPieces = [
  [
    ["","","",""],
    ["","","",""],
    ["","","d","s"],
    ["","","e","i"]
  ],
  [
    ["","","h",""],
    ["","","e",""],
    ["","","w","l"],
    ["","","",""]
  ],
  [
    ["","","",""],
    ["","","",""],
    ["l","c","r",""],
    ["","o","",""]
  ],
];

// Mock the nested components to avoid snapshotting components that are tested separately
jest.mock('../PieceDraggableView', () => 'PieceDraggableView');
jest.mock('../PieceDrawLetterGrid', () => 'PieceDrawLetterGrid');
jest.mock('../MeasureView', () => 'MeasureView');

test ('DrawPieceSection renders correctly', () => {
  const drawPieceAllowDrag = render(
    <Provider store={store}>
      <DrawPieceSection pieces={testPieces} allowDrag={true} />);
    </Provider>
  );

  drawPieceAllowDrag.getAllByType(MeasureView).forEach((measureView) => {
    fireEvent(measureView, "onMeasure", 5, 5, 100, 100, 105, 105);
  });

  const reduxPieceLocations = store.getState().gameDisplay.pieceLocations;

  const expectedLocation = {
    x:5,
    y:5,
    width:100,
    height:100,
    pageX:105,
    pageY:105,
    piece: testPieces[1],
  };

  expect(reduxPieceLocations[1]).toMatchObject(expectedLocation);
  expect(drawPieceAllowDrag.toJSON()).toMatchSnapshot();

  drawPieceAllowDrag.unmount();

  const drawPieceNoDrag = render(
    <Provider store={store}>
      <DrawPieceSection pieces={testPieces} allowDrag={false} />);
    </Provider>
  );

  drawPieceNoDrag.getAllByType(MeasureView).forEach((measureView) => {
    fireEvent(measureView, "onMeasure", 5, 5, 100, 100, 105, 105);
  });

  expect(drawPieceNoDrag).toMatchSnapshot();
  drawPieceAllowDrag.unmount();
});