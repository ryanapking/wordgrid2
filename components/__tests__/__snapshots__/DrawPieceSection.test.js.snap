// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawPieceSection renders correctly 1`] = `
Array [
  <View
    style={
      Array [
        undefined,
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "margin": 2,
          "maxHeight": "100%",
          "maxWidth": "100%",
        }
      }
    >
      <MeasureView
        onMeasure={[Function]}
        style={
          Object {
            "aspectRatio": 1,
            "backgroundColor": "gray",
            "maxHeight": "100%",
            "maxWidth": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "margin": 2,
          "maxHeight": "100%",
          "maxWidth": "100%",
        }
      }
    >
      <MeasureView
        onMeasure={[Function]}
        style={
          Object {
            "aspectRatio": 1,
            "backgroundColor": "gray",
            "maxHeight": "100%",
            "maxWidth": "100%",
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "margin": 2,
          "maxHeight": "100%",
          "maxWidth": "100%",
        }
      }
    >
      <MeasureView
        onMeasure={[Function]}
        style={
          Object {
            "aspectRatio": 1,
            "backgroundColor": "gray",
            "maxHeight": "100%",
            "maxWidth": "100%",
          }
        }
      />
    </View>
  </View>,
  ");",
]
`;

exports[`DrawPieceSection renders correctly 2`] = `
Array [
  <View
    style={
      Array [
        undefined,
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "row",
          "height": "100%",
          "justifyContent": "center",
          "width": "100%",
        },
      ]
    }
  >
    <View
      style={
        Object {
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "margin": 2,
          "maxHeight": "100%",
          "maxWidth": "100%",
        }
      }
    >
      <MeasureView
        onMeasure={[Function]}
        style={
          Object {
            "aspectRatio": 1,
            "backgroundColor": "gray",
            "maxHeight": "100%",
            "maxWidth": "100%",
          }
        }
      >
        <PieceDraggableView
          allowDrag={false}
          baseSize={100}
          piece={
            Array [
              Array [
                "",
                "",
                "",
                "",
              ],
              Array [
                "",
                "",
                "",
                "",
              ],
              Array [
                "",
                "",
                "d",
                "s",
              ],
              Array [
                "",
                "",
                "e",
                "i",
              ],
            ]
          }
          pieceIndex={0}
          style={
            Object {
              "aspectRatio": 1,
              "backgroundColor": "gray",
              "maxHeight": "100%",
              "maxWidth": "100%",
            }
          }
        >
          <PieceDrawLetterGrid
            piece={
              Array [
                Array [
                  "",
                  "",
                  "",
                  "",
                ],
                Array [
                  "",
                  "",
                  "",
                  "",
                ],
                Array [
                  "",
                  "",
                  "d",
                  "s",
                ],
                Array [
                  "",
                  "",
                  "e",
                  "i",
                ],
              ]
            }
            pieceSize={100}
          />
        </PieceDraggableView>
      </MeasureView>
    </View>
    <View
      style={
        Object {
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "margin": 2,
          "maxHeight": "100%",
          "maxWidth": "100%",
        }
      }
    >
      <MeasureView
        onMeasure={[Function]}
        style={
          Object {
            "aspectRatio": 1,
            "backgroundColor": "gray",
            "maxHeight": "100%",
            "maxWidth": "100%",
          }
        }
      >
        <PieceDraggableView
          allowDrag={false}
          baseSize={100}
          piece={
            Array [
              Array [
                "",
                "",
                "h",
                "",
              ],
              Array [
                "",
                "",
                "e",
                "",
              ],
              Array [
                "",
                "",
                "w",
                "l",
              ],
              Array [
                "",
                "",
                "",
                "",
              ],
            ]
          }
          pieceIndex={1}
          style={
            Object {
              "aspectRatio": 1,
              "backgroundColor": "gray",
              "maxHeight": "100%",
              "maxWidth": "100%",
            }
          }
        >
          <PieceDrawLetterGrid
            piece={
              Array [
                Array [
                  "",
                  "",
                  "h",
                  "",
                ],
                Array [
                  "",
                  "",
                  "e",
                  "",
                ],
                Array [
                  "",
                  "",
                  "w",
                  "l",
                ],
                Array [
                  "",
                  "",
                  "",
                  "",
                ],
              ]
            }
            pieceSize={100}
          />
        </PieceDraggableView>
      </MeasureView>
    </View>
    <View
      style={
        Object {
          "display": "flex",
          "flex": 1,
          "flexDirection": "column",
          "justifyContent": "center",
          "margin": 2,
          "maxHeight": "100%",
          "maxWidth": "100%",
        }
      }
    >
      <MeasureView
        onMeasure={[Function]}
        style={
          Object {
            "aspectRatio": 1,
            "backgroundColor": "gray",
            "maxHeight": "100%",
            "maxWidth": "100%",
          }
        }
      >
        <PieceDraggableView
          allowDrag={false}
          baseSize={100}
          piece={
            Array [
              Array [
                "",
                "",
                "",
                "",
              ],
              Array [
                "",
                "",
                "",
                "",
              ],
              Array [
                "l",
                "c",
                "r",
                "",
              ],
              Array [
                "",
                "o",
                "",
                "",
              ],
            ]
          }
          pieceIndex={2}
          style={
            Object {
              "aspectRatio": 1,
              "backgroundColor": "gray",
              "maxHeight": "100%",
              "maxWidth": "100%",
            }
          }
        >
          <PieceDrawLetterGrid
            piece={
              Array [
                Array [
                  "",
                  "",
                  "",
                  "",
                ],
                Array [
                  "",
                  "",
                  "",
                  "",
                ],
                Array [
                  "l",
                  "c",
                  "r",
                  "",
                ],
                Array [
                  "",
                  "o",
                  "",
                  "",
                ],
              ]
            }
            pieceSize={100}
          />
        </PieceDraggableView>
      </MeasureView>
    </View>
  </View>,
  ");",
]
`;
